### PR TITLE
feat/pydantic: Use Pydantic library for types

### PR DESCRIPTION
**Pull request by marioevz**
## 🗒️ Description

Simplify code by using [Pydantic](https://docs.pydantic.dev/2.6/) library.

#### src/ethereum_test_tools/common/base_types.py

A pydantic schema was defined as `ToStringSchema` which all base types inherit in order to be properly serializable.

#### src/ethereum_test_tools/common/types.py

The following types have been transformed into pydantic base models:
- Storage
- Account
- Alloc
- Withdrawal
- Environment
- Transaction
- AccessList

Types Storage and Alloc were defined as `RootModel` types due to the nature of them containing non-fixed keys in their dictionaries.

Type generics were used to simplify the creation of inherited types where the only change was the number encoding, such as `WithdrawalGeneric`, `EnvironmentGeneric` and `TransactionGeneric`, which have been used to create `FixtureWithdrawal`, `FixtureEnvironment`, and `FixtureTransaction`, respectively.

Many type methods have been refactored into cached properties, such as the transaction's RLP, in order to reduce the number of times this same value has to be computed.

Models `TransactionReceipt`,  `RejectedTransaction`, `Result`, `TransitionToolOutput`, are types that are used to parse the responses from the transition tool.

#### src/ethereum_test_tools/spec/state/types.py, src/ethereum_test_tools/spec/state/state_test.py

State test types and its fixture types have been refactored to use pydantic models.

#### src/ethereum_test_tools/spec/blockchain/types.py

Blockchain test types and its fixture types have been refactored to use pydantic models.

### TODOs

There is still room for improvement but this can be left for to subsequent PRs:
- Properly define type schemas (Add examples) and try to use pydantic primitive types to speed up the test generation.
- Refactor the types into a different library to be able to reuse common types in the `evm_transition_tool` library.

## 🔗 Related Issues
None

## ✅ Checklist
- [x] All: Set appropriate labels for the changes.
- [x] All: Considered squashing commits to improve commit history.
- [x] All: Added an entry to [CHANGELOG.md](/ethereum/execution-spec-tests/blob/main/docs/CHANGELOG.md).
- [ ] All: Considered updating the online docs in the [./docs/](/ethereum/execution-spec-tests/blob/main/docs/) directory.
- [x] Tests: All converted JSON/YML tests from [ethereum/tests](/ethereum/tests) have been added to [converted-ethereum-tests.txt](/ethereum/execution-spec-tests/blob/main/converted-ethereum-tests.txt).
- [x] Tests: Included the type and version of evm t8n tool used to locally execute test cases:  e.g., ref with commit hash or geth 1.13.1-stable-3f40e65.
- [ ] Tests: Ran `mkdocs serve` locally and verified the auto-generated docs for new tests in the [Test Case Reference](https://ethereum.github.io/execution-spec-tests/main/tests/) are correctly formatted.

### CODE REVIEW

**Comment by marioevz** [comment]
1100+ less lines of code overall.

**Review by danceratopz — APPROVED** [review]
I love it. The code is much more readable. The removal of all the `json_encoder` defs in types.py is such a joy.

The suggestions below are mainly a few additions to docstrings/comments.

File: tests/shanghai/eip4895_withdrawals/test_withdrawals.py [hunk 1]
```diff
@@ -668,9 +663,16 @@ def test_zero_amount(
         withdrawals = all_withdrawals
         post = all_post
     elif test_case == ZeroAmountTestCases.FOUR_ONE_WITH_MAX_REVERSED:
-        withdrawals = all_withdrawals
-        withdrawals.reverse()
-        set_withdrawal_index(withdrawals)
+        for i in range(len(all_withdrawals)):
+            w = all_withdrawals[len(all_withdrawals) - i - 1]
```
[the remark this answers is not in the document]
Comment (marioevz) [reply tests/shanghai/eip4895_withdrawals/test_withdrawals.py:667]: Much cleaner, thanks!

File: docs/CHANGELOG.md [hunk 2]
```diff
@@ -22,6 +23,10 @@ Test fixtures for use by clients are available for each release on the [Github r
 - 🐞 Fix CI by using Golang 1.21 in Github Actions to build geth ([#484](https://github.com/ethereum/execution-spec-tests/pull/484)).
 - 💥 "Merge" has been renamed to "Paris" in the "network" field of the Blockchain tests, and in the "post" field of the State tests ([#480](https://github.com/ethereum/execution-spec-tests/pull/480)).
 - ✨ Port entry point scripts to use [click](https://click.palletsprojects.com) and add tests ([#483](https://github.com/ethereum/execution-spec-tests/pull/483)).
+- 💥 As part of the pydantic conversion, the fixtures have the following (possibly breaking) changes ([#486](https://github.com/ethereum/execution-spec-tests/pull/486)):
+  - `expectException` fields now print the exceptions in sorted order.
```
[the remark this answers is not in the document]
Comment (marioevz) [reply docs/CHANGELOG.md:27]: Ah this is actually no longer true on the latest commit, I need to remove this.

The only reason I removed it, and also the duplicate check, is that we basically gain nothing from the fixtures being like this, and the annotated type is much cleaner than a custom list class.

**Pull request merged**